File: code/python/pq/compare_book.py
from book import Book
from functools import cmp_to_key

if __name__ == '__main__':
    books = [Book("Gone with the wind", 89, "Margaret Mitchell"),
             Book("Data structures", 120, "Unknown"),
             Book("The old man and the sea", 36, "Ernest Hemingway")]

    books.sort()

    for a in books:
        print(a)
    print('---')

    books.sort(key=lambda book: book.price)

    for a in books:
        print(a)
    print('---')

    def book_title(b):
        return b.name

    books.sort(key=book_title)

    for a in books:
        print(a)
    print('---')

    def book_title_len(b1: Book, b2: Book):
        """the longer first"""
        if len(b1.name) < len(b2.name):
            return 1
        elif len(b1.name) > len(b2.name):
            return -1
        else:
            return 0

    books.sort(key=cmp_to_key(book_title_len))

    for a in books:
        print(a)

File: code/java/pq/src/main/java/org/swufe/datastructures/CompareBook.java
package org.swufe.datastructures;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

class BookNameComparator implements Comparator<Book> {
    @Override
    public int compare(Book o1, Book o2) {
        return o1.getName().compareTo(o2.getName());
    }
}

public class CompareBook {
    public static void main(String[] args) {
        List<Book> books = new ArrayList<>();
        books.add(new Book("Gone with the wind", 89, "Margaret Mitchell"));
        books.add(new Book("Data structures", 120, "Unknown"));
        books.add(new Book("The old man and the sea", 36, "Ernest Hemingway"));
        Collections.sort(books);
        books.forEach(System.out::println);
        System.out.println("----");

        // you can use either one
        Collections.sort(books, new BookNameComparator());
        books.forEach(System.out::println);
        System.out.println("----");
        books.sort(new BookNameComparator());
        books.forEach(System.out::println);
        System.out.println("----");

        // you can use either one
        books.sort((a, b) -> a.getAuthor().compareTo(b.getAuthor()));
        books.forEach(System.out::println);
        System.out.println("----");
        books.sort(Comparator.comparing(Book::getAuthor));
        books.forEach(System.out::println);
        System.out.println("----");
    }
}
